interpreter/src/main/scala/pl/touk/nussknacker/engine/spel/parser/NuTemplateAwareExpressionParser.java

```diff
@@ -0,0 +1,287 @@
+	 * the complete evaluated string. Unwrapping is only done of the outermost delimiters
+	 * found, so the string 'hello ${foo${abc}}' would break into the pieces 'hello ' and
+	 * 'foo${abc}'. This means that expression languages that used ${..} as part of their
+	 * functionality are supported without any problem. The parsing is aware of the
+	 * structure of an embedded expression. It assumes that parentheses '(', square
+	 * brackets '[' and curly brackets '}' must be in pairs within the expression unless
+	 * they are within a string literal and a string literal starts and terminates with a
+	 * single quote '.
+	 * @param expressionString the expression string
+	 * @return the parsed expressions
+	 * @throws ParseException when the expressions cannot be parsed
+	 */
+	private ExpressionWithPosition[] parseExpressions(String expressionString, ParserContext context) throws ParseException {
+		List<ExpressionWithPosition> expressions = new ArrayList<>();
+		String prefix = context.getExpressionPrefix();
+		String suffix = context.getExpressionSuffix();
+		int startIdx = 0;
+
+		while (startIdx < expressionString.length()) {
+			int prefixIndex = expressionString.indexOf(prefix, startIdx);
+			if (prefixIndex >= startIdx) {
+				// an inner expression was found - this is a composite
+				if (prefixIndex > startIdx) {
+					expressions.add(new ExpressionWithPosition(new LiteralExpression(expressionString.substring(startIdx, prefixIndex)), startIdx, prefixIndex));
+				}
+				int afterPrefixIndex = prefixIndex + prefix.length();
+				int suffixIndex = skipToCorrectEndSuffix(suffix, expressionString, afterPrefixIndex);
+				if (suffixIndex == -1) {
+					throw new ParseException(expressionString, prefixIndex,
+							"No ending suffix '" + suffix + "' for expression starting at character " +
+							prefixIndex + ": " + expressionString.substring(prefixIndex));
+				}
+				if (suffixIndex == afterPrefixIndex) {
+					throw new ParseException(expressionString, prefixIndex,
+							"No expression defined within delimiter '" + prefix + suffix +
+							"' at character " + prefixIndex);
+				}
+				String expr = expressionString.substring(prefixIndex + prefix.length(), suffixIndex);
+				// expr.trim() replaced with this code to calculate actual start and end (without leading/trailing spaces)
+				String exprStripLeading = expr.stripLeading();
+				String exprStripTrailing = exprStripLeading.stripTrailing();
+				int leadingWhitespaces = expr.length() - exprStripLeading.length();
+				int trailingWhitespaces = exprStripLeading.length() - exprStripTrailing.length();
+				if (expr.isEmpty()) {
+					throw new ParseException(expressionString, prefixIndex,
+							"No expression defined within delimiter '" + prefix + suffix +
+							"' at character " + prefixIndex);
+				}
+				expressions.add(new ExpressionWithPosition(doParseExpression(expr),prefixIndex + prefix.length() + leadingWhitespaces, suffixIndex + trailingWhitespaces));
```

Comment:
1. Why do you keep in the `ExpressionWithPosition` class parsed expression with whitespace but start index after whitespace stripping? Would've been a difference if you had passed `expr.trim()` into `doParseExpression` method?
2. Why `suffixIndex + trailingWhitespaces`? Shouldn't it be minus?

Reply:
1. Good point, this works now because in test the expression is longer than spaces before. Added another case with more spaces and it fails. Fixed that.
2. Another good point, fixed.